### PR TITLE
Correctly handle multi-threaded queries involving dates

### LINKED ISSUE #290: JDBC: Multiple, simultaneous queries including dates may return spurious results
When many queries are run at the same time & those queries return dates, these dates can get mixed during the return process.

### PR DESCRIPTION
**Pull request by fhcarter**
Fixes #290 

Code was using `SimpleDateFormat` in an unsafe manner.  This has been corrected.

Added new test to verify things.  Runs 150 simultaneous threads.  Ramped manually up higher, no change, but no reason to keep that in.  Failure without fix happened with only a few threads.

Did other code hygiene whilst there.

### CODE REVIEW

File: jdbcSource/src/main/java/io/vantiq/extsrc/jdbcSource/JDBC.java [hunk 1]
```diff
@@ -219,18 +216,24 @@ public int processPublish(String sqlQuery) throws VantiqSQLException {
...
+    Map[] createMapFromResults(ResultSet queryResults) throws VantiqSQLException {
+        ArrayList<HashMap<String, Object>> rows = new ArrayList<HashMap<String, Object>>();
         try {
             if (!queryResults.next()) { 
-                return rows.toArray(new HashMap[rows.size()]);
+                return rows.toArray(new HashMap[0]);
             } else {
                 ResultSetMetaData md = queryResults.getMetaData(); 
                 int columns = md.getColumnCount();
+
```
Comment (fhcarter) [inline jdbcSource/src/main/java/io/vantiq/extsrc/jdbcSource/JDBC.java:228]: The code below is the actual fix.  Remainder is just code cleanup

File: jdbcSource/src/main/java/io/vantiq/extsrc/jdbcSource/JDBCCore.java [hunk 2]
```diff
@@ -10,12 +10,10 @@
```
Comment (fhcarter) [inline jdbcSource/src/main/java/io/vantiq/extsrc/jdbcSource/JDBCCore.java:10]: Changes here pretty much entirely code cleanup

File: jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java [hunk 3]
```diff
@@ -8,31 +8,41 @@
...
 
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 import static org.junit.Assume.assumeFalse;
 import static org.junit.Assume.assumeTrue;
 import static org.junit.Assert.assertEquals;
 
+import com.google.gson.JsonArray;
```
Comment (fhcarter) [inline jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java:19]: Moved imports around to match standard.

File: jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java [hunk 4]
```diff
@@ -54,12 +64,50 @@
...
             + "testTime TIME, testDec decimal(5,2));";
+
     static final String PUBLISH_QUERY_EXTENDED_TYPES = "INSERT INTO TestTypes VALUES (1, '" + TIMESTAMP + "', '" + DATE + "',"
             + " '" + TIME + "', 145.86);";
     static final String SELECT_QUERY_EXTENDED_TYPES = "SELECT * FROM TestTypes;";
     static final String DELETE_ROW_EXTENDED_TYPES = "DELETE FROM TestTypes;";
     static final String DELETE_TABLE_EXTENDED_TYPES = "DROP TABLE TestTypes;";
-    
+
+    static final Integer TS_COUNT = 25;
```
Comment (fhcarter) [inline jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java:74]: Test constants to test heavy date usage in multithreaded cases.

File: jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java [hunk 5]
```diff
@@ -141,15 +189,15 @@ public void setup() {
...
         vantiq.setAccessToken(testAuthToken);
     }
-    
-    /* Uncomment this BeforeClass when debugging to ensure tables have been properly deleted.
-    
+
     @BeforeClass
-    public static void debugSetup() throws VantiqSQLException {
-        tearDown();
+    public static void setupEnv() throws IOException {
```
Comment (fhcarter) [inline jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java:194]: Handle failing tests that hadn't been changed after config changes.  Problem with too many setup variables needed....

File: jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java [hunk 6]
```diff
@@ -235,6 +283,13 @@ public static void tearDown() throws VantiqSQLException {
             } catch (VantiqSQLException e) {
                 // Shouldn't throw Exception
             }
+
+            // Delete parallel Dates table
```
Comment (fhcarter) [inline jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java:287]: cleanup as required...

File: jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java [hunk 7]
```diff
@@ -386,6 +444,104 @@ public void testExtendedTypes() throws VantiqSQLException {
...
+            e.printStackTrace();
+            fail("Should not have thrown exception." + e);
+        }
+
+        // Insert values into the newly created table
+        // Here, we will set up a bunch of sets of 25 rows, where each row has a bunch of dates in it.
+
+        ArrayList<Map<String, Object>> dataMap = new ArrayList<>();
+        try {
+            for (int i = 0; i < 25; i++) {
```
Comment (fhcarter) [inline jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java:469]: create tables full of dates that vary by id...

File: jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java [hunk 8]
```diff
@@ -386,6 +444,104 @@ public void testExtendedTypes() throws VantiqSQLException {
...
+
+        // Now, to verify our parallel operations, we'll fire off a bunch of threads all doing the same
+        // query across this table full of dates.  We should get everything with correct results,
+        // but we had had the date results scrambled due to use of non-threadsafe methods.
+
+        AtomicReference<Exception> threadFailed = new AtomicReference<>();
+        Runnable querier = () -> {
+            // Select the values from the table and make sure the data is retrieved correctly
+            try {
+                Map[] qres = jdbc.processQuery(SELECT_QUERY_MANY_DATES);
```
Comment (fhcarter) [inline jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java:490]: In each thread, read the entire table of dates & make sure it makes sense.  If any exceptions are thrown, save them for the master test to detect.

File: jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java [hunk 9]
```diff
@@ -386,6 +444,104 @@ public void testExtendedTypes() throws VantiqSQLException {
...
+                        String timeTest = (String) row.get("testTime");
+                        assert timeTest.matches(timePattern);
+                        assert timeTest.equals(FORMATTED_TIME);
+                    }
+                }
+            } catch (Exception e) {
+                threadFailed.set(e);
+            }
+        };
+
```
Comment (fhcarter) [inline jdbcSource/src/test/java/io/vantiq/extsrc/jdbcSource/TestJDBC.java:516]: Now fire off a bunch of threads doing all this stuff...

**Comment by fhcarter** [comment]
@AlexBlumer -- With our new absence, can you take a quick look at this...

**Review by AlexBlumer — APPROVED** [review]
LGTM

**Comment by fhcarter** [comment]
[Jenkins Run](https://jenkins.vantiq.com/jenkins/job/vantiq-extension-sources-branch/38/)

**Pull request merged**